shell/platform/android/platform_view_android_jni.cc

```diff
@@ -757,52 +758,311 @@ bool PlatformViewAndroid::Register(JNIEnv* env) {
+    return false;
+  }
+
+  g_handle_platform_message_method =
+          env->GetMethodID(g_flutter_engine_class->obj(),
+                           "handlePlatformMessage", "(Ljava/lang/String;[BI)V");
+
+  if (g_handle_platform_message_method == nullptr) {
+    FML_LOG(ERROR) << "Could not locate handlePlatformMessage method";
+    return false;
+  }
+
+  g_handle_platform_message_response_method =
+          env->GetMethodID(g_flutter_engine_class->obj(),
+                           "handlePlatformMessageResponse", "(I[B)V");
+
+  if (g_handle_platform_message_response_method == nullptr) {
+    FML_LOG(ERROR) << "Could not locate handlePlatformMessageResponse method";
+    return false;
+  }
+
+  g_update_semantics_method =
+      env->GetMethodID(g_flutter_engine_class->obj(), "updateSemantics",
+                       "(Ljava/nio/ByteBuffer;[Ljava/lang/String;)V");
+
+  if (g_update_semantics_method == nullptr) {
+    FML_LOG(ERROR) << "Could not locate updateSemantics method";
+    return false;
+  }
+
+  g_update_custom_accessibility_actions_method = env->GetMethodID(
+      g_flutter_engine_class->obj(), "updateCustomAccessibilityActions",
+      "(Ljava/nio/ByteBuffer;[Ljava/lang/String;)V");
+
+  if (g_update_custom_accessibility_actions_method == nullptr) {
+    FML_LOG(ERROR) << "Could not locate updateCustomAccessibilityActions method";
+    return false;
+  }
+
+  g_on_first_frame_method = env->GetMethodID(g_flutter_engine_class->obj(),
+                                             "onFirstFrame", "()V");
+
+  if (g_on_first_frame_method == nullptr) {
+    FML_LOG(ERROR) << "Could not locate onFirstFrame method";
+    return false;
+  }
+
+  g_on_engine_restart_method = env->GetMethodID(
+      g_flutter_engine_class->obj(), "onPreEngineRestart", "()V");
+
+  if (g_on_engine_restart_method == nullptr) {
+    FML_LOG(ERROR) << "Could not locate onEngineRestart method";
+    return false;
+  }
+
+  return true;
+}
+
+bool PlatformViewAndroid::Register(JNIEnv* env) {
+  FML_LOG(ERROR) << "Registering";
+  if (env == nullptr) {
+    FML_LOG(ERROR) << "No JNIEnv provided";
+    return false;
+  }
+
+  g_flutter_callback_info_class = new fml::jni::ScopedJavaGlobalRef<jclass>(
+      env, env->FindClass("io/flutter/view/FlutterCallbackInformation"));
+  if (g_flutter_callback_info_class->is_null()) {
+    FML_LOG(ERROR) << "Could not locate FlutterCallbackInformation class";
+    return false;
+  }
+
+  g_flutter_callback_info_constructor = env->GetMethodID(
+      g_flutter_callback_info_class->obj(), "<init>",
+      "(Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;)V");
+  if (g_flutter_callback_info_constructor == nullptr) {
+    FML_LOG(ERROR) << "Could not locate FlutterCallbackInformation constructor";
+    return false;
+  }
+
+  // TODO(mattcarroll): this assumes the use of the new API is based on what we're compiling against but that's not true. Need to support both simultaneously.
```

Comment:
Does this mean we need to update build files to generate different flutter.jar's?

At the moment this .cc file contains support for the old way of doing things, and the new one.  My TODO comment was pointing out that the compile-time existence of the new Java class doesn't imply that the developer is necessarily using the new Java class.  So the TODO says we need explicit instruction from the Java side as to which universe we're supporting.

I'm not sure off the top of my head how to support customized JARs because developers pull a single build of the engine with the framework, right?  How would a developer choose one JAR or the other in that scenario?  

Or am I totally misunderstanding your question?

Reply:
Does the flutter.jar built on your branch include both the old and new Java classes?  Or does it only include the new classes?

If the former, then the JNI code should register implementations for both the old and new method signatures.  If the latter, then you can remove the registration code for the old classes.